mm2 param handling

elmir-ma opened the pull request:
Generalized parameter handling for mapping flags. It started as parameter handling for minimap2 

1) Adding --junc-bed option (handled by converting gtf to bed12 at mapping.smk which is inputted to minimap2 mapping rule).
2) adding sr flag for short-read samples. First created a mapping strategy just for mm2 short reads in map_strategy.py but proved to be too bulky. Solution by Marvin: elegant config.yaml structure for parameter handling for mappers (below in the comments)

elmir-ma committed:
mm2 short-reads mapping strategy

elmir-ma committed:
final log symlink is mapper dependentnow

elmir-ma committed:
mm2 specific final-log symlink handlingg

mschilli87 commented:
Also see https://github.com/rajewsky-lab/spacemake/pull/128.

marvin-jens commented:
Hi Elmir,
many thanks for opening a PR! It is the way. :)
I see no problems with your changes. However, I am somewhat uncomfortable to add `mm2_sr` as a new "mapper". As you maybe saw from my out-commented code on the `wrap_minimap2` branch, I pondered this idea before and decided not to pursue further. Two reasons:

  * dislike growing `if, elif, elif, elif...` blocks. What if sb needs a new switch for STAR? Add yet another pseudo-mapper?
  * a general mechanism to pass **sample-specific parameters to the mapper** would be nice.
  
I would therefore invite a bit further discussion. Rough idea:

Have a section in the config YAML like this:
```
mapping:
  bowtie2:
    default:
        - "--local"
        - " -L 10 -D 30 -R 30 -k 10"
        - "--ignore-quals"
        - "--score-min=L,0,1.5"  # require 75% of perfect match (2=base match)
  STAR:
    default: 
        - "--outSAMprimaryFlag AllBestScore"
        - "--outSAMattributes Standard"
        - "--outSAMunmapped Within"
        - "--outStd BAM_Unsorted"
        - "--outSAMtype BAM Unsorted"
        - "--limitOutSJcollapsed 5000000"
    
  mm2:
    default: 
       - "-x spliced"
    sr: 
       - "-x spliced:sr"
```
Combined with another modifier (how about the humble dot `.`?) in the `map_strategy`: `mm2.sr:genome`.

upsides:

  * separate config from code
  * document default parameter choices explicitly
  * allow full customization
  * scale better. Less code, more YAML as needed
  
downsides:

  * need to add a bit more logic (but very similar to what we already have for barcode and adapter-flavor) to fetch stuff from config.yaml at the appropriate places in `map_strategy.py` and perhaps `mapping.smk`.
  * ???

I have a feeling that this may ultimately lead to code-simplifications.

nukappa commented:
Thanks a lot @elmir-ma . I agree with @marvin-jens 's take on adding the params inside the config.yaml (more robust, reproducible, documentation etc).

elmir-ma commented:
Thank you all for very useful comments, I had a feeling I made map_strategy.py  unnecessarily more crowded .  Config tweaking sounds very reasonable. Still developing intuition of the pipeline

elmir-ma committed:
Revert mm2_sr mapper approach, keep improved log handling

elmir-ma committed:
Add --junc-bed support for minimap2 mapping

- Add create_junc_bed rule to convert GTF to BED12 using paftools.js
- Update map_reads_mm2 to require junction BED file as input
- Pass --junc-bed flag to minimap2 for improved splice junction detection
- BED file is automatically generated from annotation when mm2 mapping is used

elmir-ma committed:
Add --junc-bed flag to mm2 that would improve junction mapping. It is dependent on the minimap2 map strategy and conditional on existence of reference annotation (in case reference is without annotation).

marvin-jens commented on spacemake/snakemake/mapping.smk:
```diff
@@ -115,6 +114,9 @@ def get_map_inputs(wc, mapper="STAR"):
     if hasattr(mr, "ann_final"):
         d['annotation'] = mr.ann_final
+        # For minimap2, also add junction BED file when annotation exists
+        if mapper == "mm2" and mr.ann_final:
```
Hi Elmir, I'm not a fan of inserting paths into the code here. Original design was to keep the folder structure and filename patterns in `spacemake/snakemake/variables.py`.

marvin-jens replied:
One more thought. The junc_bed file belongs to the annotation. checking for mr.ann_final is not necessarily correct. For example, if you map to two separate mm2 indices in a row, only the last one will get junc_bed. Better to check if you actually have annotation

elmir-ma committed:
Move junc-bed path to variables.py and check ann_path instead of ann_final (conceptualy cleaner)

elmir-ma committed:
Use species_reference_junc_bed variable in create_junc_bed rule

marvin-jens reviewed (approved):
I have reviewed all changes and it looks very solid. Merging
